tensorboard/defs/internal/js.bzl

```diff
@@ -0,0 +1,146 @@
+
+load("@build_bazel_rules_nodejs//:providers.bzl", "JSNamedModuleInfo", "NpmPackageInfo", "node_modules_aspect")
+
+def _tf_dev_js_binary_impl(ctx):
+    files_depsets = []
+
+    bootstrap_and_deps = ctx.attr._ambient_deps + ctx.attr.deps
+    for dep in bootstrap_and_deps:
+        if JSNamedModuleInfo in dep:
+            # Collect UMD modules compiled by tf_ts_library
+            files_depsets.append(dep[JSNamedModuleInfo].sources)
+        elif not NpmPackageInfo in dep and hasattr(dep, "files"):
+            # Collect Bazel's File or UMD modules from a MPM package
+            files_depsets.append(dep.files)
+
+    for target in ctx.attr._anonymous_umd_deps:
+        file = target.files.to_list()[0]
+        module_name = ctx.attr._anonymous_umd_deps[target]
+        named_file = ctx.actions.declare_file(file.path + ".named_umd.js")
+
+        # Patch anonymous umd modules to have named in their declarations. For instance,
+        # it converts `define(['exports'], ...` to `define('d3', ['exports'], ...`.
+        # `define`'s argument behaves differently based on arity. For instance:
+        # 1: expects the argument to be a factory function to be invoked. Anonymous and
+        #    no dependency.
+        # 2: expects an array then a function. First arguments define dependencies to be
+        #    injected itno the factory. Anonymous with dependencies.
+        # 3: expects string, an array, then, a function. First argument is name of the
+        #    module. Named module with deps.
+        ctx.actions.expand_template(
+            template = file,
+            output = named_file,
+            substitutions = {
+                # d3 and three
+                "define(['exports']": "define('%s', ['exports']" % module_name,
+                # Lodash
+                "define(function()": "define('%s', function()" % module_name,
+                # Zone.js
+                "define(factory": "define('%s', factory" % module_name,
+            },
+            is_executable = False,
+        )
+        files_depsets.append(depset([named_file]))
+
+    files = depset(transitive = files_depsets)
+
+    # files can contain package.json that is not even runnable under require.js. Prune it
+    # out.
+    js_files = [
+        f
+        for f in files.to_list()
+        if f.path.endswith(".js")
+    ]
+
+    ctx.actions.write(
+        output = ctx.outputs.manifest,
+        content = "\n".join([file.path for file in js_files]),
+        is_executable = False,
+    )
+
+    entry_point_module_name = ctx.workspace_name + "/" + _remove_ext(ctx.file.entry_point.short_path)
+
+    concat_command = ";".join(
+        [
+            "awk 'BEGINFILE {print \"// file: \"FILENAME}{print}' * " + " ".join([file.path for file in js_files]) + " >" + ctx.outputs.js.path,
+            # `require` the entry module name so it is evaluated.
+            "echo ';require([\"%s\"]);' >>" % entry_point_module_name + ctx.outputs.js.path,
+        ],
+    )
```

Comment:
Can we use shell arguments rather than manually interpolating (analogous
to using parameters in SQL queries)? Could write your command as:

```python
    concat_command = """
        output="$1" && shift
        entry_point="$1" && shift
        {
            awk 'BEGINFILE { print "// file: " FILENAME } { print }' "$@"
            printf ';require(["%s"]);\n' "${entry_point}"
        } >"${output}"
    """
    concat_args = (
        [ctx.outputs.js.path, entry_point_module_name]
        + [file.path for file in js_files]
    )
    ctx.actions.run_shell(..., arguments = concat_args)
```